src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/Html/DefaultHtmlCodeActionProvider.cs

```diff
@@ -0,0 +1,72 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.LanguageServer.CodeActions.Models;
+using Microsoft.AspNetCore.Razor.LanguageServer.Common;
+using Microsoft.AspNetCore.Razor.LanguageServer.Extensions;
+using Microsoft.AspNetCore.Razor.LanguageServer.Formatting;
+using Microsoft.CodeAnalysis.Razor.Workspaces.Extensions;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.CodeActions
+{
+    internal class DefaultHtmlCodeActionProvider : HtmlCodeActionProvider
+    {
+        private readonly RazorDocumentMappingService _documentMappingService;
+
+        public DefaultHtmlCodeActionProvider(RazorDocumentMappingService documentMappingService)
+        {
+            _documentMappingService = documentMappingService;
+        }
+
+        public override async Task<IReadOnlyList<RazorVSInternalCodeAction>?> ProvideAsync(
+            RazorCodeActionContext context,
+            IEnumerable<RazorVSInternalCodeAction> codeActions,
+            CancellationToken cancellationToken)
+        {
+            var results = new List<RazorVSInternalCodeAction>();
+
+            foreach (var codeAction in codeActions)
+            {
+                if (codeAction.Edit is not null)
+                {
+                    await RemapeAndFixHtmlCodeActionEditAsync(_documentMappingService, context.CodeDocument, codeAction, cancellationToken).ConfigureAwait(false);
+
+                    results.Add(codeAction);
```

Comment:
I'd be tempted to use `yield return` here but I can't say that it would get you much other than "#pranavpoints". If you have arguments for/against it I'd be interested but it's mostly a curiosity for me.

Reply:
The endpoint fires off all providers as tasks, and waits for them all to complete before combining. I don't know if that is actually gaining anything, but making this an iterator would remove whatever parallelism might exist